docs/proposals/202105256-cluster-class-and-managed-topologies.md

```diff
@@ -468,17 +468,90 @@ This section talks about updating a cluster which was created using a `ClusterCl
+- it is also important to notice that:
+    - ClusterClass and managed topologies can work **only** with control plane providers implementing support for the `spec.version` field;
+      Additionally, it is required to provide support for the `status.version` field reporting the minimum
+      API server version in the cluster; the control plane contract will be extended accordingly.
+    - ClusterClass and managed topologies can work both with control plane providers implementing support for
+      machine infrastructures and with control plane providers not supporting this feature.
+      The control plane contract will be extended introducing well know fields where the machine template
+      should be defined (in case this feature is supported).
+    - ClusterClass and managed topologies can work both with control plane providers implementing support for
+      `spec.replicas` and with control plane provider not supporting this feature.
+
+**Impact on the infrastructure providers**:
+
+- the provider implementers are required to implement the InfrastructureClusterTemplate type (e.g. `AWSClusterTemplate`, `AzureClusterTemplate` etc.).
+
+#### Conventions for template types implementation
+
+Given that it is required to implement new templates, let's remind the conventions used for
+defining templates and the corresponding objects:
+
+Templates:
+
+- Template fields must match or be a subset of the corresponding generated object.
+- A template can't accept values which are not valid for the corresponding generated object,
+  otherwise creating an object derived from a template will fail.
+  
+Objects generated from the template:
+
+- For the fields existing both in the object and in the corresponding template:
+    - The object can't have additional validation rules than the template,
+      otherwise creating an object derived from a template could fail.
+- For the fields existing only in the object but not in the corresponding template:
+    - Fields must be optional or a default value must be automatically assigned,
+      otherwise creating an object derived from a template will fail.
+
+**Note:** The existing InfrastructureMachineTemplate and BootstrapMachineTemplate objects already
+comply those conventions via explicit rules implemented in the code or via operational practices
+(otherwise creating machines will not be working already today).
+
+**Note:** As per this proposal, the definition of ClusterClass is immutable. The CC definition consists 
+of infrastructure object references, say AWSMachineTemplate, which could be immutable. For such immutable
+infrastructure objects, hard-coding the image identifiers leads to those templates being tied to a particular
+kubernetes version, thus making kubernetes version upgrades impossible. Hence, when using CC, infrastructure
+objects MUST NOT have mandatory static fields whose values prohibit version upgrades.
```

Comment:
I wonder if we should phrase it a bit differently. 

I think the main problem is:

We don't want InfrastructureMachineTemplates to be tied to specific Kubernetes version, because otherwise we would have to update the ref to a InfrastructureMachineTemplate (to another one which supports the new Kubernetes version) in the ClusterClass **at the same time** when we update the Kubernetes version in all Clusters which are using the ClusterClass, which is impossible. Thus we need Kubernetes versions to be independent of the templates.

tl;dr compared to the current text. The immutability isn't the biggest problem, the main problem is that we would have to change the ref to a new template and the Kubernetes version both at the same time

I think this argumentation holds now and when we make the ClusterClass mutable. Of course both are correct.

Hope that all makes sense, feel free to keep it as is or modify it a bit :)

Reply:
This is a copy paste from the orginal proposal and in this PR I'm just reshuffling and but not changing this bits.
Given that this is going to be addressed when implementing mutability transformations/patches in the next iteration and the note will be removed soon, I would like to avoid to change a phrasing we all agreed upon.